### PR TITLE
Add support for regex quantifiers

### LINKED ISSUE #109: Support for quantifiers and the shuffle operator in regular expressions
Is there a specific reason why quantifiers for regular expressions are not supported? (so things like `a{2,10}`) If there is no specific reason, would you be open to a PR that implements it?

Also, what do you think of adding a special regex shuffle operator? This would be very handy to express languages like "all permutations of a,b,c". My proposal would be to use the `~` character and implement a separate parser for that

### PR DESCRIPTION
**Pull request by eliotwrobson**
Pull request for quantifiers, resolves #109

### CODE REVIEW

**Comment by eliotwrobson** [comment]
@caleb531 Not _quite_ done yet, but an initial version of the quantifier syntax is working. The main obstacle I'm running into right now has to do with the way that the default input symbols are retrieved. The issue is that to create the factory function for the wildcard, the lexer needs to have all of the non-reserved symbols that appear in a given regex.

However, with the default method of retrieving the input symbols, numbers used in the quantifiers will get picked up even though they don't appear as literals in the given regex. Because of the difference between where the factory function is defined and called, this creates a potential sharp edge that may cause issues for some people. Do you have any thoughts on this? There are a couple of ways to get around that but I'm not quite sure which one would be best.

File: automata/regex/parser.py [hunk 1]
```diff
@@ -140,9 +140,9 @@ def intersection(self, other):
             if epsilon_transitions_b is not None:
                 state_dict = new_transitions.setdefault(curr_state_name, {})
                 state_dict.setdefault('', set()).update(
-                    map(get_state_name, product([q_a], epsilon_transitions_b))
+                    map(get_state_name, zip(repeat(q_a), epsilon_transitions_b))
```
Comment (caleb531) [inline automata/regex/parser.py:143]: @eliotwrobson What's the advantage of using `repeat(q_a)` over `[q_a]`?

File: automata/regex/regex.py [hunk 2]
```diff
@@ -33,7 +34,7 @@ def isequal(re1, re2, *, input_symbols=None):
     nfa1 = NFA.from_regex(re1, input_symbols=input_symbols)
     nfa2 = NFA.from_regex(re2, input_symbols=input_symbols)
 
-    return nfa1 == nfa2
+    return DFA.from_nfa(nfa1) == DFA.from_nfa(nfa2)
```
Comment (caleb531) [inline automata/regex/regex.py]: @eliotwrobson I thought we got the NFA equality algorithm working correctly. Is that not actually the case (thus leading you to compare the DFAs instead)?

Comment (eliotwrobson) [reply automata/regex/regex.py]: Oh no, it was working fine but I was passing in the wrong NFA during testing and tried a bunch of random stuff before realizing it was my mistake 🤣 this got committed by accident. I think this change is reverted in the current version though.

File: automata/regex/parser.py (hunk 1 above)
Comment (eliotwrobson) [reply automata/regex/parser.py:143]: So I saw someone do this in an itertools video and mainly wanted to try it out, but after some timing turns out its slower... Another lesson on why it's not good to try and be too clever 😅

File: automata/regex/regex.py (hunk 2 above)
Comment (caleb531) [reply automata/regex/regex.py]: @eliotwrobson Gotcha. I can confirm it's reverted. Thanks for the clarity!

File: automata/regex/parser.py (hunk 1 above)
Comment (caleb531) [reply automata/regex/parser.py:143]: @eliotwrobson Well, how many iterations are those benchmarks based off of? I am curious if one solution scales better than the other.

Comment (eliotwrobson) [reply automata/regex/parser.py:143]: This was 1 million. I'm actually curious now, let me try some different numbers.

**Comment by eliotwrobson** [comment]
@caleb531 looking at this again, how do you feel about changing the default input symbols for creating a regex to something very general, like all alphas + numbers? I think it removes the awkwardness of having the default regex to NFA conversion be only over the symbols in present literals. If not, I think there's a workaround to the wildcard creation issue that keeps this same default, just makes the creation code a little messier.

**Comment by eliotwrobson** [comment]
@caleb531 going to go ahead and flip this to open now that I had the chance to clean it up. The only controversial thing here is changing the default alphabet used for parsing regex, but I think this actually makes it easier if someone is just playing around with regexes. Performance/memory usage shouldn't be an issue since people can still manually set the input symbols.

File: automata/regex/parser.py (hunk 1 above)
Comment (caleb531) [reply automata/regex/parser.py]: @eliotwrobson Just checking on the status of your benchmarking, since this thread is the last unresolved conversation on this PR.

Comment (eliotwrobson) [reply automata/regex/parser.py]: The changes discussed in this thread actually got merged in #130, but here is more detailed benchmarking:

image

Even at the smaller sizes, the difference is somewhat noticeable.

Comment (caleb531) [reply automata/regex/parser.py]: @eliotwrobson Oh okay great! I'm satisfied then lol. Will resolve this thread and review the rest of your PR in preparation to approve/merge.

File: automata/regex/parser.py [hunk 3]
```diff
@@ -181,40 +183,65 @@ def concatenate(self, other):
 
         self._final_states = other._final_states
 
-    def kleene_star(self):
+    def repeat(self, lower_bound, upper_bound):
```
Comment (caleb531) [inline automata/regex/parser.py:186]: @eliotwrobson So I understand the introduction of the `repeat` method, but don't we still need to keep the `kleene_star` method lest we break API backwards-compatibility?

**Comment by caleb531** [comment]
> Haven't updated docs completely, will do once this gets closer to being merged.

@eliotwrobson Just reviewed this PR again—everything looks good. Did you still have any documentation updates to make before I merge?

**Comment by eliotwrobson** [comment]
@caleb531 nope! I added the docs updates to the PR already (just noting the quantifier inclusion and changing the default alphabet), so I think this is good to go. If we feel like there's more docs needed, that can be done before this goes on main.

**Comment by caleb531** [comment]
@eliotwrobson Perfect! Will approve and merge now, then. Thank you for your work on this functionality!

**Review by caleb531 — APPROVED** [review]
